qiskit/pulse/reference_manager.py

```diff
@@ -0,0 +1,57 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Management of schedule block references."""
+
+from typing import Tuple
+from qiskit.pulse.exceptions import PulseError
+
+
+class ReferenceManager(dict):
+    """Dictionary wrapper to manage pulse schedule references."""
+
+    def unassigned(self) -> Tuple[Tuple[str, ...], ...]:
+        """Get the keys of unassigned references.
+
+        Returns:
+            Tuple of reference keys.
+        """
+        keys = []
+        for key, value in self.items():
+            if value is None:
+                keys.append(key)
+        return tuple(keys)
+
+    def __setitem__(self, key, value):
+        if key in self and self[key] is not None:
+            # Check subroutine conflict.
+            if self[key] != value:
+                raise PulseError(
+                    f"Subroutine {key} is already assigned to the reference of the current scope, "
+                    "however, the newly assigned schedule conflicts with the existing schedule. "
+                    "This operation was not successfully done."
+                )
+            return
+        super().__setitem__(key, value)
```

Comment:
This isn't generally sufficient when subclassing `dict`:
```python
>>> class MyDict(dict):
...     def __setitem__(self, key, value):
...             print("here!")
...             return super().__setitem__(key, value)
...
>>> d = MyDict()
>>> d["hello"] = "world"
here!
>>> d.update({"a": "b"})
>>> # nothing printed!
```

The Python built-ins generally implement all their methods separately in C space, and don't call the Python-space overrides.  The quickest fix is just to derive from `collections.UserDict` instead, which will do what you expect.

Reply:
Thanks, done in https://github.com/Qiskit/qiskit-terra/pull/8005/commits/ae506e849ed3a93ce74a5c6f0ba910ee7c6fb2a7